api/src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -103,6 +115,16 @@ public final boolean isRunning() {
      * Returns a read-only view of parents.
      */
     public List<FlowNode> getParents() {
+        if (parents == null) {
+            parents = new ArrayList<FlowNode>(parentIds.size());
+            for (String id : parentIds) {
```

Comment:
This code is thread-unsafe. It's a part of the  public API, hence :bug: . Use a local variable or synchronize the method